### PR TITLE
fix(iast): ast patching side effect

### PR DESCRIPTION
**Pull request by gnufede**
IAST: This fix resolves an issue where, at AST patching to replace code with IAST aspects, the original function/method was passed as an extra parameter for accurate patching (i.e. check the function inside the aspect), and this strategy lead to unintentionally triggers side effects in methods obtained from an expression (like ``decode`` in ``file.read(n).decode()``), resulting in unexpected multiple calls to the expression (``file.read(n)`` in the example).

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/appsec/_iast/_taint_tracking/aspects.py [hunk 1]
```diff
@@ -158,16 +169,19 @@ def slice_aspect(candidate_text, start, stop, step) -> Any:
...
-    # type: (Callable, int, Any, Any) -> Any
-    if not isinstance(orig_function, BuiltinFunctionType):
+    # type: (Optional[Callable], int, Any, Any) -> Any
+    if orig_function and not isinstance(orig_function, BuiltinFunctionType):
         if flag_added_args > 0:
             args = args[flag_added_args:]
         return orig_function(*args, **kwargs)
 
+    if len(args) < 2:
+        return args[0].extend(*args[1:], **kwargs)
```
Comment (avara1986) [inline ddtrace/appsec/_iast/_taint_tracking/aspects.py:179]: Could you please add a comment and/or example that explains why we need this logic? If not, we may forget the rationale behind it in a month :sweat_smile:

File: ddtrace/appsec/_iast/_taint_tracking/aspects.py [hunk 2]
```diff
@@ -498,7 +519,11 @@ def decode_aspect(orig_function, flag_added_args, *args, **kwargs):
         self = args[0]
         args = args[flag_added_args:]
     else:
-        return orig_function(*args, **kwargs)
+        if orig_function:
+            return orig_function(*args, **kwargs)
+        self = args[0]
+        args = args[1:]
```
Comment (avara1986) [inline ddtrace/appsec/_iast/_taint_tracking/aspects.py:525]: ditto

Comment (avara1986) [inline ddtrace/appsec/_iast/_taint_tracking/aspects.py:525]: ditto

File: ddtrace/appsec/_iast/_taint_tracking/aspects.py (hunk 1 above)
Comment (gnufede) [reply ddtrace/appsec/_iast/_taint_tracking/aspects.py:179]:
```suggestion
        # If we're not receiving at least 2 arguments, means the call was
        # ``x.extend()`` and not ``x.extend(y)``
        # so either not the extend we're looking for, or no changes in taint ranges.
        return args[0].extend(*args[1:], **kwargs)
```

**Review by avara1986 — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Comment by avara1986** [comment]
LGTM, great Job @gnufede !

**Pull request merged**